README.md
# react-bootstrap-daterangepicker

[![NPM version](https://badge.fury.io/js/react-bootstrap-daterangepicker.svg)](http://badge.fury.io/js/react-bootstrap-daterangepicker)
[![Build Status](https://travis-ci.org/skratchdot/react-bootstrap-daterangepicker.svg?branch=master)](https://travis-ci.org/skratchdot/react-bootstrap-daterangepicker)
[![Code Climate](https://codeclimate.com/github/skratchdot/react-bootstrap-daterangepicker.png)](https://codeclimate.com/github/skratchdot/react-bootstrap-daterangepicker)
[![Coverage Status](https://coveralls.io/repos/skratchdot/react-bootstrap-daterangepicker/badge.svg?branch=master&service=github)](https://coveralls.io/github/skratchdot/react-bootstrap-daterangepicker?branch=master)

[![NPM](https://nodei.co/npm/react-bootstrap-daterangepicker.png)](https://npmjs.org/package/react-bootstrap-daterangepicker)

# 🚨 Deprecation Notice 🚨

> I put this project on github because I used it briefly for a project back in 2014. I haven't used it for years, and have recommended
> looking for a "pure react" date picker library. I might continue to merge small PRs, but will not be giving this library much/any
> support. I recommend using one of the [other react date picker](#other-react-date-pickers) libraries listed below.

## Description

A date/time picker for react (using bootstrap). This is a react wrapper around
an existing jQuery/bootstrap library (it is not a pure react port):

[bootstrap-daterangepicker](https://github.com/dangrossman/bootstrap-daterangepicker)

## Getting Started

1.  Install the needed peer dependencies:
    `npm install --save bootstrap-daterangepicker react jquery moment`

2.  Install the module with:
    `npm install --save react-bootstrap-daterangepicker`

3.  Include the bootstrap@4 css and fonts in your project.
    (aka `import 'bootstrap/dist/css/bootstrap.css';`)

4.  Include the bootstrap-daterangepicker css in your project.
    (aka `import 'bootstrap-daterangepicker/daterangepicker.css';`)

5.  This is a commonjs library. You will need a tool like browserify/webpack/etc to build your code.

```javascript
import React, { Component } from 'react';
import DateRangePicker from 'react-bootstrap-daterangepicker';
// you will need the css that comes with bootstrap@3. if you are using
// a tool like webpack, you can do the following:
import 'bootstrap/dist/css/bootstrap.css';
// you will also need the css that comes with bootstrap-daterangepicker
import 'bootstrap-daterangepicker/daterangepicker.css';

class MyComponent {
  render() {
    return (
      <DateRangePicker
        initialSettings={{ startDate: '1/1/2014', endDate: '3/1/2014' }}
      >
        <button>Click Me To Open Picker!</button>
      </DateRangePicker>
    );
  }
}
```

## Documentation

For in depth documentation, see the original
[bootstrap-daterangepicker](https://github.com/dangrossman/bootstrap-daterangepicker) project page.

You can pass all the settings from the original plugin to the `initialSettings` prop:

- **&lt;input&gt;, alwaysShowCalendars, applyButtonClasses, applyClass,
  autoApply, autoUpdateInput, buttonClasses, cancelButtonClasses, cancelClass,
  dateLimit, drops, endDate, isCustomDate, isInvalidDate, linkedCalendars,
  locale, maxDate, maxSpan, maxYear, minDate, minYear, moment, opens, parentEl,
  ranges, showCustomRangeLabel, showDropdowns, showISOWeekNumbers,
  showWeekNumbers, singleDatePicker, startDate, template, timePicker,
  timePicker24Hour, timePickerIncrement, timePickerSeconds**

You can listen to the following 8 events:

- **onShow**: `callback(event, picker)` thrown when the widget is shown
- **onHide**: `callback(event, picker)` thrown when the widget is hidden
- **onShowCalendar**: `callback(event, picker)` thrown when the calendar is shown
- **onHideCalendar**: `callback(event, picker)` thrown when the calendar is hidden
- **onApply**: `callback(event, picker)` thrown when the apply button is clicked
- **onCancel**: `callback(event, picker)` thrown when the cancel button is clicked
- **onEvent**: `callback(event, picker)` thrown when any of the 6 events above are triggered
- **onCallback**: `callback(start, end, label)` thrown when the start/end dates change

You MUST pass a single child element to the `<DateRangePicker />` component- and it MUST be a DOM element.
Passing custom react components is not currently supported b/c this lib needs a single dom node to initialize.

NOTE: This component should be used as an [Uncontrolled Component](https://reactjs.org/docs/uncontrolled-components.html). If you try
to control the value of your child `<input />`, then you will probably encounter issues.

There are 2 methods from the upstream lib that can be called: `setStartDate` and `setEndDate`, but you need to use refs when doing so.
Please view the storybook for an example of this.

### Examples

For more usage examples, please view the storybook:  
https://projects.skratchdot.com/react-bootstrap-daterangepicker/

#### Simple button example

```javascript
<DateRangePicker>
  <button type="button" className="btn btn-primary">
    click to open
  </button>
</DateRangePicker>
```

#### Simple input example

```javascript
<DateRangePicker>
  <input type="text" className="form-control" />
</DateRangePicker>
```

#### Initialize with a startDate and endDate

```javascript
<DateRangePicker
  initialSettings={{ startDate: '01/01/2020', endDate: '01/15/2020' }}
>
  <input type="text" className="form-control" />
</DateRangePicker>
```

#### Example event handler:

```javascript
class SomeReactComponent extends React.Component {
  handleEvent(event, picker) {
    console.log(picker.startDate);
  }
  handleCallback(start, end, label) {
    console.log(start, end, label);
  }
  render() {
    return (
      <DateRangePicker onEvent={this.handleEvent} onCallback={this.handleCallback}>
        <input />
      </DateRangePicker>;
  }
}
```

## Release Notes

Release notes can be found in the
[Changelog](https://github.com/skratchdot/react-bootstrap-daterangepicker/blob/master/CHANGELOG.md).

## Links

- [Source Code](https://github.com/skratchdot/react-bootstrap-daterangepicker)
- [Changelog](https://github.com/skratchdot/react-bootstrap-daterangepicker/blob/master/CHANGELOG.md)
- [Live Demo](http://projects.skratchdot.com/react-bootstrap-daterangepicker/)
- [Original Plugin](https://github.com/dangrossman/bootstrap-daterangepicker)

## Other React Date Pickers

- [react-date-range](https://github.com/Adphorus/react-date-range)
- [react-dates](https://github.com/airbnb/react-dates)
- [react-datepicker](https://github.com/Hacker0x01/react-datepicker)
- [react-datetimerange-picker](https://github.com/wojtekmaj/react-datetimerange-picker)

**NOTE: Please submit a PR if there are other date pickers you can recommend**

## License

Copyright (c) 2014 skratchdot  
Uses the original [bootstrap-daterangepicker](https://github.com/dangrossman/bootstrap-daterangepicker) license.


CHANGELOG.md
# Changelog

## [8.0.0](https://github.com/skratchdot/react-bootstrap-daterangepicker/tree/8.0.0) (2022-08-24)

[Full Changelog](https://github.com/skratchdot/react-bootstrap-daterangepicker/compare/7.0.0...8.0.0)

**Closed issues:**

- Change default active class [\#246](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/246)
- singleDatePicker Attribute Not Working [\#240](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/240)
- parentEl in initialSettings - How to use [\#239](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/239)
- How to prevent user from not selecting date more than a week [\#238](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/238)
- Get name of predefined range as a prop [\#236](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/236)
- I'm trying to use the showCustomRangeLabel option as true but it has no effect [\#234](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/234)
- How to set max date [\#233](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/233)
- Date range picker input box doesn't show up calendar icon ? [\#229](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/229)
- How to pass maxspan of 31 days for allowing max range of 31 days between start and end dates? [\#227](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/227)
- How do i allow apply on only selecting start date? [\#226](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/226)
- TS definitions broken since the `initialSettings` change. [\#224](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/224)

**Merged pull requests:**

- react 18 fixes [\#257](https://github.com/skratchdot/react-bootstrap-daterangepicker/pull/257) ([danhab99](https://github.com/danhab99))
- updating dependencies [\#237](https://github.com/skratchdot/react-bootstrap-daterangepicker/pull/237) ([skratchdot](https://github.com/skratchdot))

## [7.0.0](https://github.com/skratchdot/react-bootstrap-daterangepicker/tree/7.0.0) (2020-08-12)

[Full Changelog](https://github.com/skratchdot/react-bootstrap-daterangepicker/compare/7.0.0-beta.1...7.0.0)

## [7.0.0-beta.1](https://github.com/skratchdot/react-bootstrap-daterangepicker/tree/7.0.0-beta.1) (2020-08-11)

[Full Changelog](https://github.com/skratchdot/react-bootstrap-daterangepicker/compare/7.0.0-beta.0...7.0.0-beta.1)

## [7.0.0-beta.0](https://github.com/skratchdot/react-bootstrap-daterangepicker/tree/7.0.0-beta.0) (2020-08-11)

[Full Changelog](https://github.com/skratchdot/react-bootstrap-daterangepicker/compare/6.0.0...7.0.0-beta.0)

**Closed issues:**

- Specifying the language [\#218](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/218)
- singleDatePicker - Change date from input. [\#217](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/217)
- singleDatePicker shows range of dates when date is changed dynamically [\#211](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/211)
- end date active error [\#210](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/210)
- accessibility support or keyboard navigation is not supported [\#207](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/207)
- how can i put checkbox inside calender [\#205](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/205)
- Calendar jumps to start date after selecting end date in range selection [\#202](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/202)
- this.props undefined [\#201](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/201)
- V4 is unusable with inputs [\#200](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/200)
- showDropdown/timePicker failing [\#199](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/199)
- How to set date range only for startDate [\#198](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/198)
- Customise Range selection [\#197](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/197)
- Calendar opens and closes on each input event [\#195](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/195)
- startDate should be always set before endDate [\#194](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/194)
- The NPM package does not match the Github code [\#193](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/193)
- How can I customise the calendar? [\#192](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/192)
- Property applyButtonClasses, cancelButtonClasses not apply [\#191](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/191)
- autoUpdateInput - Unknown props `autoUpdateInput`, `singleDatePicker`, `startDate` on \<div\> tag. Remove these props from the element. [\#190](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/190)
- Uncaught ReferenceError: DateRangePicker is not defined [\#189](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/189)
- minDate is not working [\#187](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/187)
- Can you please update react-bootstrap-daterangepicker to react 16 , we are in dire need of it, thanks a mil [\#186](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/186)
- How do I manually enter a date using the daterangepicker@3.0.3? [\#185](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/185)
- endDate dose not work. [\#184](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/184)
- this.$picker.daterangepicker is not a function error while running Jest test cases [\#183](https://github.com/skratchdot/react-bootstrap-daterangepicker/issues/183)

stories/index.js
import React, { useState, useRef } from 'react';
import jQuery from 'jquery';
import moment from 'moment';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, text } from '@storybook/addon-knobs';
import DateRangePicker from '../src/index.js';
import 'bootstrap/dist/css/bootstrap.css';
import 'bootstrap-daterangepicker/daterangepicker.css';

// expose jQuery to window for debugging
window.jQuery = window.$ = jQuery;

storiesOf('DateRangePicker', module)
  .addDecorator((story, context) => withInfo()(story)(context))
  .addDecorator((story) => <div style={{ margin: 40 }}>{story()}</div>)
  .addDecorator(withKnobs)
  .add('simple button', () => {
    const buttonLabel = text('label', 'click to open');
    return (
      <DateRangePicker>
        <button type="button" className="btn btn-primary">
          {buttonLabel}
        </button>
      </DateRangePicker>
    );
  })
  .add('simple input', () => {
    return (
      <DateRangePicker>
        <input type="text" className="form-control col-4" />
      </DateRangePicker>
    );
  })
  .add('with start and end date', () => {
    const startDate = text('startDate', '1/1/2017');
    const endDate = text('endDate', '1/5/2017');
    return (
      <DateRangePicker initialSettings={{ startDate, endDate }}>
        <input type="text" className="form-control col-4" />
      </DateRangePicker>
    );
  })
  .add('log events', () => {
    return (
      <DateRangePicker
        onApply={action('onApply')}
        onCancel={action('onCancel')}
        onHide={action('onHide')}
        onHideCalendar={action('onHideCalendar')}
        onShow={action('onShow')}
        onShowCalendar={action('onShowCalendar')}
        onEvent={action('onEvent')}
        onCallback={action('onCallback')}
      >
        <input type="text" className="form-control col-4" />
      </DateRangePicker>
    );
  })
  .add('date picker with dropdowns', () => {
    return (
      <DateRangePicker initialSettings={{ showDropdowns: true }}>
        <input type="text" className="form-control col-4" />
      </DateRangePicker>
    );
  })
  .add('single date picker', () => {
    return (
      <DateRangePicker
        initialSettings={{
          singleDatePicker: true,
          showDropdowns: true,
          startDate: '10/18/1984',
          minYear: 1901,
          maxYear: parseInt(moment().format('YYYY'), 10),
        }}
        onCallback={(start) => {
          const years = moment().diff(start, 'years');
          alert('You are ' + years + ' years old!');
        }}
      >
        <input type="text" className="form-control col-4" />
      </DateRangePicker>
    );
  })
  .add('date range picker with times', () => {
    return (
      <DateRangePicker
        initialSettings={{
          timePicker: true,
          startDate: moment().startOf('hour').toDate(),
          endDate: moment().startOf('hour').add(32, 'hour').toDate(),
          locale: {
            format: 'M/DD hh:mm A',
          },
        }}
      >
        <input type="text" className="form-control col-4" />
      </DateRangePicker>
    );
  })
  .add('predefined date ranges', () => {
    const [state, setState] = useState({
      start: moment().subtract(29, 'days'),
      end: moment(),
    });
    const { start, end } = state;
    const handleCallback = (start, end) => {
      setState({ start, end });
    };
    const label =
      start.format('MMMM D, YYYY') + ' - ' + end.format('MMMM D, YYYY');
    return (
      <DateRangePicker
        initialSettings={{
          startDate: start.toDate(),
          endDate: end.toDate(),
          ranges: {
            Today: [moment().toDate(), moment().toDate()],
            Yesterday: [
              moment().subtract(1, 'days').toDate(),
              moment().subtract(1, 'days').toDate(),
            ],
            'Last 7 Days': [
              moment().subtract(6, 'days').toDate(),
              moment().toDate(),
            ],
            'Last 30 Days': [
              moment().subtract(29, 'days').toDate(),
              moment().toDate(),
            ],
            'This Month': [
              moment().startOf('month').toDate(),
              moment().endOf('month').toDate(),
            ],
            'Last Month': [
              moment().subtract(1, 'month').startOf('month').toDate(),
              moment().subtract(1, 'month').endOf('month').toDate(),
            ],
          },
        }}
        onCallback={handleCallback}
      >
        <div
          id="reportrange"
          className="col-4"
          style={{
            background: '#fff',
            cursor: 'pointer',
            padding: '5px 10px',
            border: '1px solid #ccc',
            width: '100%',
          }}
        >
          <i className="fa fa-calendar"></i>&nbsp;
          <span>{label}</span> <i className="fa fa-caret-down"></i>
        </div>
      </DateRangePicker>
    );
  })
  .add('input initially empty', () => {
    const handleApply = (event, picker) => {
      picker.element.val(
        picker.startDate.format('MM/DD/YYYY') +
          ' - ' +
          picker.endDate.format('MM/DD/YYYY')
      );
    };
    const handleCancel = (event, picker) => {
      picker.element.val('');
    };
    return (
      <DateRangePicker
        initialSettings={{
          autoUpdateInput: false,
          locale: {
            cancelLabel: 'Clear',
          },
        }}
        onApply={handleApply}
        onCancel={handleCancel}
      >
        <input type="text" className="form-control col-4" defaultValue="" />
      </DateRangePicker>
    );
  })
  .add('set start date from "outside" the component', () => {
    const myRef = useRef();
    const changeStartDate = () => {
      myRef.current.setStartDate(moment().subtract(1, 'week'));
    };
    return (
      <div>
        <DateRangePicker ref={myRef}>
          <input type="text" className="form-control col-4" />
        </DateRangePicker>
        <br />
        <button onClick={changeStartDate}>set startDate to 1 week ago</button>
      </div>
    );
  })
  .add('use onCallback to display start/end dates', () => {
    const [state, setState] = useState({});
    const handleCallback = (start, end) => {
      setState({ start, end });
    };
    return (
      <>
        <DateRangePicker onCallback={handleCallback}>
          <input type="text" className="form-control col-4" />
        </DateRangePicker>
        <br />
        <pre>state: {JSON.stringify(state)}</pre>
      </>
    );
  })
  .add('change initialSettings (range data)', () => {
    const StoryComp = () => {
      const changeButtonLabel = text(
        'change button label',
        'change range label'
      );
      const keyRef = useRef(Date.now());
      const [dates, setDates] = useState({
        startDate: moment('2020/03/01'),
        endDate: moment('2020/03/15'),
      });
      const [ranges, setRanges] = useState({
        ['First Range']: [
          moment().subtract(2, 'days'),
          moment().add(2, 'days'),
        ],
      });
      const handleApply = (event, picker) => {
        setDates({
          startDate: picker.startDate,
          endDate: picker.endDate,
        });
      };
      const randomNumber = () => Math.floor(Math.random() * 20) + 1;
      const handleChangeRanges = () => {
        keyRef.current = Date.now();
        setRanges({
          [`Range ${Date.now()}`]: [
            moment().subtract(randomNumber(), 'days').startOf('day'),
            moment().add(randomNumber(), 'days').startOf('day'),
          ],
        });
      };
      return (
        <div>
          <DateRangePicker
            key={keyRef.current}
            onApply={handleApply}
            onCancel={action('onCancel')}
            onEvent={action('onEvent')}
            onHide={action('onHide')}
            onHideCalendar={action('onHideCalendar')}
            onShow={action('onShow')}
            onShowCalendar={action('onShowCalendar')}
            initialSettings={{ ranges }}
          >
            <input type="text" className="form-control col-4" />
          </DateRangePicker>
          <br />
          <h4>
            startDate: <small>{dates.startDate.format()}</small>
          </h4>
          <h4>
            endDate: <small>{dates.endDate.format()}</small>
          </h4>
          <h4>
            ranges: <small>{JSON.stringify(ranges)}</small>
          </h4>
          <button onClick={handleChangeRanges}>{changeButtonLabel}</button>
        </div>
      );
    };
    return (
      <div>
        <StoryComp />
      </div>
    );
  });
